### PR TITLE
Support for Llama-3_1-Nemotron-51B

### PR DESCRIPTION
**Pull request by ymcki**
*Make sure to read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

More details is here:
https://github.com/ggerganov/llama.cpp/issues/10648

Seems like my changes in vocab.py doesn't really break CI test. 

It might be a better idea to not to modify vocab.py but instead ask the user to fix the tokenizer_config.json instead. In that case, you can ignore the changes I made in vocab.py.

### CODE REVIEW

**Comment by bartowski1182** [comment]
I wonder if it's also a better idea not to group this with the normal llama archs since it requires so many changes, may be better to make it its own model type?

**Comment by ymcki** [comment]
> I wonder if it's also a better idea not to group this with the normal llama archs since it requires so many changes, may be better to make it its own model type?

I think src/llama.cpp doesn't change that much but convert_hf_to_gguf.py does have more changes. Anyway, I can make another fork to make it a separate model type and submit another pull request.

What do you think about the vocab.py problem? Should I just leave the original vocab.py as is and ask people to fix tokenizer_config.json instead?

**Comment by ymcki** [comment]
Created a separate Deci Model. This version doesn't change vocab.py and relies on people manually fixing 51B model's tokenizer_config.json.

**Comment by ymcki** [comment]
Any updates?

File: src/llama.cpp [hunk 1]
```diff
@@ -10934,6 +11033,166 @@ struct llm_build_context {
...
+
+        const float kq_scale = hparams.f_attention_scale == 0.0f ? 1.0f/sqrtf(float(n_embd_head)) : hparams.f_attention_scale;
+        for (int il = 0; il < n_layer; ++il) {
+            struct ggml_tensor * inpSA = inpL;
+            const int64_t n_head_kv = hparams.n_head_kv(il);
+            const int64_t n_head    = hparams.n_head(il);
+
+            if (n_head == 0) // attention-free layer of Llama-3_1-Nemotron-51B
+                cur = inpL;
+            else {
```
Comment (ggerganov) [inline src/llama.cpp:11065]:
```suggestion
            if (n_head == 0) { // attention-free layer of Llama-3_1-Nemotron-51B
                cur = inpL;
            } else {
```

File: src/llama.cpp [hunk 2]
```diff
@@ -10934,6 +11033,166 @@ struct llm_build_context {
...
+                        LLM_NORM_RMS, cb, il);
+                cb(cur, "attn_norm", il);
+            }
+
+            if (n_head > 0 && n_head_kv == 0) { // "linear attention" of Llama-3_1-Nemotron-51B
+                cur = llm_build_lora_mm(lctx, ctx0, model.layers[il].wo, cur);
+                cb(cur, "wo", il);
+            } else if (n_head > 0)
+            // self-attention
+            {
```
Comment (ggerganov) [inline src/llama.cpp:11078]:
```suggestion
            } else if (n_head > 0) {
                // self-attention
```

**Review by ggerganov — APPROVED** [review]

**Comment by ngxson** [comment]
To fix editorconfig / flake8 tests, you need to modify your source code to remove trailing spaces / add new line.

And to fix server CI, you need to merge latest commits from master branch

**Comment by ymcki** [comment]
Can someone approve the workflows?

**Comment by ymcki** [comment]
Yay! Finally passed all checks! :)

**Review by slaren — APPROVED** [review]

File: src/llama.cpp [hunk 3]
```diff
@@ -11304,6 +11403,165 @@ struct llm_build_context {
...
+                        model.layers[il].attn_norm, NULL,
+                        LLM_NORM_RMS, cb, il);
+                cb(cur, "attn_norm", il);
+            }
+
+            if (n_head > 0 && n_head_kv == 0) { // "linear attention" of Llama-3_1-Nemotron-51B
+                cur = llm_build_lora_mm(lctx, ctx0, model.layers[il].wo, cur);
+                cb(cur, "wo", il);
+            } else if (n_head > 0) {
+            // self-attention
```
Comment (slaren) [inline src/llama.cpp:11447]:
```suggestion
                // self-attention
```

File: src/llama.cpp [hunk 4]
```diff
@@ -11304,6 +11403,165 @@ struct llm_build_context {
...
+        // KQ_mask (mask for 1 head, it will be broadcasted to all heads)
+        struct ggml_tensor * KQ_mask = build_inp_KQ_mask();
+
+        const float kq_scale = hparams.f_attention_scale == 0.0f ? 1.0f/sqrtf(float(n_embd_head)) : hparams.f_attention_scale;
+        for (int il = 0; il < n_layer; ++il) {
+            struct ggml_tensor * inpSA = inpL;
+            const int64_t n_head_kv = hparams.n_head_kv(il);
+            const int64_t n_head    = hparams.n_head(il);
+
+            if (n_head == 0) { // attention-free layer of Llama-3_1-Nemotron-51B
```
Comment (slaren) [inline src/llama.cpp:11433]:
```suggestion
            if (n_head == 0) {
                // attention-free layer of Llama-3_1-Nemotron-51B
```

File: src/llama.cpp [hunk 5]
```diff
@@ -11304,6 +11403,165 @@ struct llm_build_context {
...
+                cur = inpL;
+            } else {
+                // norm
+                cur = llm_build_norm(ctx0, inpL, hparams,
+                        model.layers[il].attn_norm, NULL,
+                        LLM_NORM_RMS, cb, il);
+                cb(cur, "attn_norm", il);
+            }
+
+            if (n_head > 0 && n_head_kv == 0) { // "linear attention" of Llama-3_1-Nemotron-51B
```
Comment (slaren) [inline src/llama.cpp:11443]:
```suggestion
            if (n_head > 0 && n_head_kv == 0) {
                // "linear attention" of Llama-3_1-Nemotron-51B
```

**Pull request merged**